File: 0289-game-of-life/0289-game-of-life.cpp
class Solution {
public:
    vector<pair<int,int>> dir = {{0,1},{0,-1},{1,0},{-1,0},{1,1},{1,-1},{-1,1},{-1,-1}};
    void gameOfLife(vector<vector<int>>& board) {
        int m = board.size(), n = board[0].size(); 
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                int neighbors = 0; 
                for(pair<int,int>& p : dir){
                    int nX = i + p.first; 
                    int nY = j + p.second; 

                    if(nX >= 0 && nX < m && nY >= 0 && nY < n && (board[nX][nY] == 1 || board[nX][nY] == 3)) neighbors++; 
                }
                
                if(board[i][j] == 1){
                    if(neighbors < 2 || neighbors > 3) board[i][j] = 3; // alive → dead (encode as 3)
                }
                if(board[i][j] == 0 && neighbors == 3){
                    board[i][j] = 2; // dead → alive (encode as 2)
                }
            }
        }

        // decode states
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(board[i][j] == 2) board[i][j] = 1; // was dead, now alive
                else if(board[i][j] == 3) board[i][j] = 0; // was alive, now dead
            }
        }
    }
};

File: 0289-game-of-life/0289-game-of-life.java
class Solution {
    public void gameOfLife(int[][] board) {
        int m = board.length, n = board[0].length; 
        int[][] dir = {{0,1},{0,-1},{1,0},{-1,0},{1,1},{1,-1},{-1,1},{-1,-1}}; 


        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                
                int count = 0; 

                for(int k = 0; k < 8; k++){
                    int nX = i + dir[k][0]; 
                    int nY = j + dir[k][1]; 

                    if(nX < 0 || nY < 0 || nX >= m || nY >= n) continue; 

                    if(board[nX][nY] > 0) count++; 
                }

                if(board[i][j] == 0){
                    if(count == 3) board[i][j] = -1; 
                }
                if(board[i][j] > 0){
                    if(count < 2 || count > 3) board[i][j] = 2; 
                }
            }
        }

        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(board[i][j] == -1) board[i][j] = 1; 
                if(board[i][j] == 2) board[i][j] = 0; 
            }
        }
    }
}